src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -222,4 +240,118 @@ begin
+  rw [valuation_def, valuation.extend_to_localization_apply_map_apply],
+  exact hr,
+end
+
+/-- Uniformizers are nonzero. -/
+lemma valuation_uniformizer_ne_zero :
+  (classical.some (v.valuation_exists_uniformizer K)) ≠ 0 :=
+begin
+  have hu := (classical.some_spec (v.valuation_exists_uniformizer K)),
+  exact (valuation.ne_zero_iff _).mp (ne_of_eq_of_ne hu with_zero.coe_ne_zero),
+end
+
+/-! ### Completions with respect to adic valuations
+Given a Dedekind domain `R` with field of fractions `K` and a maximal ideal `v` of `R`, we define
+the completion of `K` with respect to its `v`-adic valuation, denoted `v.adic_completion`, and its
+ring of integers, denoted `v.adic_completion_integers`. -/
+
+variable {K}
+
+/-- `K` as a valued field with the `v`-adic valuation. -/
+def adic_valued : valued K (with_zero (multiplicative ℤ)) := valued.mk' v.valuation
+
+lemma adic_valued_def {x : K} : (v.adic_valued.v : _) x = v.valuation x := rfl
+
+/-- The uniform space structure on `K` corresponding to the `v`-adic valuation.
+This cannot be made an instance since it depends on the choice of `v`. -/
+def adic_valued_uniform_space : uniform_space K := v.adic_valued.to_uniform_space
+
+lemma adic_valued_uniform_add_group : @uniform_add_group K v.adic_valued_uniform_space _ :=
+v.adic_valued.to_uniform_add_group
+
+/-- The topological space structure on `K` corresponding to the `v`-adic valuation.
+  This cannot be made an instance since it depends on the choice of `v`. -/
+def adic_valued_topological_space : topological_space K :=
+v.adic_valued_uniform_space.to_topological_space
+
+lemma adic_valued_completable_top_field : @completable_top_field K _ v.adic_valued_uniform_space :=
+@valued.completable K _ _ _ (adic_valued v)
+
+lemma adic_valued_separated_space : @separated_space K v.adic_valued_uniform_space :=
+@valued_ring.separated K _ _ _ (adic_valued v)
+
+variables (K)
+
+/-- The completion of `K` with respect to its `v`-adic valuation. -/
+def adic_completion := @uniform_space.completion K v.adic_valued_uniform_space
+
+instance : field (v.adic_completion K) :=
+@field_completion K _ v.adic_valued_uniform_space _ _ v.adic_valued_uniform_add_group
+
+instance : inhabited (v.adic_completion K) := ⟨0⟩
+
+variables {K}
+instance valued_adic_completion : valued (v.adic_completion K) (with_zero (multiplicative ℤ)) :=
+@valued.valued_completion _ _ _ _ v.adic_valued
+
+lemma valued_adic_completion_def {x : v.adic_completion K} :
+  valued.v x = @valued.extension K _ _ _ (adic_valued v)  x := rfl
+
+instance adic_completion_complete_space : complete_space (v.adic_completion K) :=
+@uniform_space.completion.complete_space K v.adic_valued_uniform_space
+
+instance adic_completion.has_lift_t : has_lift_t K (v.adic_completion K) :=
+(infer_instance : has_lift_t K (@uniform_space.completion K v.adic_valued_uniform_space))
+
+variables (K)
+/-- The ring of integers of `adic_completion`. -/
+def adic_completion_integers : subring (v.adic_completion K) :=
```

Comment:
You could use `valuation.valuation_subring` here to produce an actual valuation subring of the completion (the carrier set is defeq with `valuation.integer`).

Reply:
Thanks!